cl/kafka: throttle shadow link produce traffic

bharathv opened the pull request:
Applies the same throttles that apply to kafka produce/consume traffic. Plumbs the snc_manager that propagates the delays into shadow link code and sleeps for those delays.

<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [x] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

* none

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv committed:
cl/mux_consumer: wire up snc_quota_mgr

Mechanical wiring up by updating all places. The integration logic
will be in the next commit.

bharathv committed:
cl/kafka: implement tput throttling for mux consumer

The fetches and subsequent produces in mux consumer are throttled using
the same kafka throughput limits via snc_quota_mgr.

The data is considered as produce requests for throttling purposes.

pgellert commented on src/v/cluster_link/service.cc:
```diff
@@ -224,12 +233,17 @@ class default_link_factory : public link_factory {
           cluster, cfg);
 
+        std::unique_ptr<kafka::snc_quota_context> quota_ctx;
+        snc_quota_mgr.get_or_create_quota_context(quota_ctx, client_id);
```
I think `get_or_create_quota_context` needs to be called inside the `mux_remote_consumer` in a `kafka_throughput_control.watch(...)` to ensure that updates to `kafka_throughput_control` are reflected.

bharathv committed:
cl/kafka: add a test for throttling

bharathv commented:
/dt

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--9369beff-7ebd-46b6-94ea-d7fab872df30-->
<!--TEST RESULTS-->
<!--6217afb0-f0a6-4720-9f79-0bf9f78822f1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73797">build#73797</a></summary>


| test_class                       | test_method                | test_arguments                                                                                                                              | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                              |
|:---------------------------------|:---------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkingReplicationTests    | test_replication_basic     | {"shuffle_leadership": false, "source_cluster_spec": {"cluster_type": "kafka", "kafka_quorum": "COMBINED_KRAFT", "kafka_version": "3.8.0"}} | integration | https://buildkite.com/redpanda/redpanda/builds/73797#0199c564-8069-474b-8019-8c20e44564de | FLAKY         | 20/21    | upstream reliability is '95.78544061302682'. current run reliability is '95.23809523809523'. drift is 0.54735 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkingReplicationTests&test_method=test_replication_basic        |
| WriteCachingFailureInjectionTest | test_unavoidable_data_loss | null                                                                                                                                        | integration | https://buildkite.com/redpanda/redpanda/builds/73797#0199c565-7301-4cdd-ae49-e04a864d7804 | FLAKY         | 18/21    | upstream reliability is '91.74174174174175'. current run reliability is '85.71428571428571'. drift is 6.02746 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionTest&test_method=test_unavoidable_data_loss |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/73923">build#73923</a></summary>


| test_class                          | test_method              | test_arguments                                                                                                                                   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                 | test_history                                                                                                                                                           |
|:------------------------------------|:-------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:---------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acad-4977-bf04-a3cfe09aeeba | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acaf-4164-b26e-eed748a17566 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acb0-479b-8c69-6d37eed76b44 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acb1-4991-bb65-cf25d6cba602 | FAIL          | 0/16     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acb3-42bb-8dc9-7278310a4b18 | FAIL          | 0/16     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": false}         | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-aca8-41a6-b3ea-ce124d45ef7b | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": false, "mixed_versions": false, "with_iceberg": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acaa-4f7e-bd74-5b2e9768ac8f | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": true}  | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acac-494a-8267-41fb16c46615 | FAIL          | 0/16     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": false, "with_iceberg": true}          | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acaf-4164-b26e-eed748a17566 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acb3-42bb-8dc9-7278310a4b18 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-aca8-41a6-b3ea-ce124d45ef7b | FAIL          | 0/1      | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-aca9-4f2b-a59b-fa295be2ffb5 | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}  | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acaa-4f7e-bd74-5b2e9768ac8f | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acab-4a32-a159-07a28426890c | FAIL          | 0/18     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}          | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acac-494a-8267-41fb16c46615 | FAIL          | 0/15     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}           | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acaf-4164-b26e-eed748a17566 | FAIL          | 0/1      |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}   | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acb1-4991-bb65-cf25d6cba602 | FAIL          | 0/14     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RandomNodeOperationsTest            | test_node_operations     | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true}           | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-aca8-41a6-b3ea-ce124d45ef7b | FAIL          | 0/1      | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations            |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false}                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc4e-929f-44b1-9b6b-6299fb72203e | FAIL          | 0/21     | The test has failed across all retries | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| RedpandaNodeOperationsSmokeTest     | test_node_ops_smoke_test | {"cloud_storage_type": 1, "mixed_versions": false}                                                                                               | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acab-4a32-a159-07a28426890c | FAIL          | 0/17     |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RedpandaNodeOperationsSmokeTest&test_method=test_node_ops_smoke_test |
| WriteCachingFailureInjectionE2ETest | test_crash_all           | {"use_transactions": false}                                                                                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/73923#0199cc50-acb1-4991-bb65-cf25d6cba602 | FLAKY         | 15/21    |                                        | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all       |
</details>
<!--/body-->

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.h:
```diff
@@ -85,7 +90,10 @@ class mux_remote_consumer {
       _partitions;
+    ss::sstring _client_id;
     std::unique_ptr<kafka::client::direct_consumer> _consumer;
+    std::unique_ptr<kafka::snc_quota_context> _snc_quota_ctx;
```
nit: don't see this used in this commit

michael-redpanda replied:
ah n/m i see how it's used in the future commit, this is fine

bharathv replied:
yes just plumbed it in this commit and used it later.

michael-redpanda commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -175,6 +199,27 @@ bool mux_remote_consumer::can_ignore_partition_data(
+        _snc_quota_mgr.get_or_create_quota_context(_snc_quota_ctx, _client_id);
+        _snc_quota_mgr.record_request_receive(
+          *_snc_quota_ctx, total_fetch_bytes);
+        auto delays = _snc_quota_mgr.get_shard_delays(*_snc_quota_ctx);
+        if (delays.request > ss::lowres_clock::duration::zero()) {
+            vlog(
+              cllog.debug,
+              "Throttling fetch processing for {}",
+              std::chrono::duration_cast<std::chrono::milliseconds>(
+                delays.request));
+            co_await ss::sleep_abortable(delays.request, _as);
+        }
```
question:  do we want to calculate an enforced value (using something akin to `throttling_state` in connection_context https://github.com/redpanda-data/redpanda/blob/80f80388ac253c010110c61fb1d92189b94d6127/src/v/kafka/server/connection_context.h#L440-L470)?  Kind of a science experiment question - would we experience worst throttling using a request vs an enforce.

bharathv replied:
My understanding is this is not needed. This logic exists in connection_context because the clients may not honor the throttling request (which is what the throttling_state tracks).  The logic there is something like "If the client didn't honor the last throttling requset, lets enforce the current throttling delay on the server by injecting a sleep".  In this case since we always honor the throttle request, we don't need that additional tracking. Anyway this was my interpretation, would be good if @pgellert can confirm too.

pgellert replied:
Yes exactly as @bharathv said. That logic is only for enforcing the delays server-side if the client didn't.

michael-redpanda reviewed (dismissed):
lgtm - just one question but imo not blocking

bharathv reviewed (commented):
I'll fix the unit test in the next push.

pgellert commented on src/v/cluster_link/replication/mux_remote_consumer.cc:
```diff
@@ -175,6 +199,27 @@ bool mux_remote_consumer::can_ignore_partition_data(
+        if (delays.request > ss::lowres_clock::duration::zero()) {
+            vlog(
+              cllog.debug,
+              "Throttling fetch processing for {}",
+              std::chrono::duration_cast<std::chrono::milliseconds>(
+                delays.request));
```
nit: add the unit (`ms`) into the log message

bharathv replied:
fmt takes care of that already..

pgellert replied:
oh of course...

pgellert reviewed (dismissed):
looks good 🚢

bharathv committed:
cl/tests: remove toggle_sessions variant in fixture tests

We donot toggle the sessions configuration in the middle of the test,
effectively it is just testing without sessions.

bharathv reviewed (commented):
force push has no core changes, deflaked the unit test.

bharathv commented:
> topic_purger_test

Failures seem unrelated.

bharathv commented:
/ci-repeat 1

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/73923-->
<!--95c37758-1566-417b-a0d1-ff8134a16ced-->
<!--e9fcd5bc-2813-4229-9133-d2f92df61507-->
<!--7036cec8-1d26-46b6-ae7a-3f0be6b1a2c9-->
<!--fab39ce5-14fc-489d-ba56-05b46ab2b3e2-->
<!--91a795d9-5b26-428d-8400-12dc9a75f76e-->
<!--title-->
## Retry command for [Build#73923](https://buildkite.com/redpanda/redpanda/builds/73923)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/random_node_operations_test.py::RedpandaNodeOperationsSmokeTest.test_node_ops_smoke_test@{"cloud_storage_type":1,"mixed_versions":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":false,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"adjacent_merge","enable_failures":true,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":false,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":1,"compaction_mode":"sliding_window","enable_failures":true,"mixed_versions":false,"with_iceberg":true}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"adjacent_merge","enable_failures":false,"mixed_versions":false,"with_iceberg":false}
tests/rptest/tests/random_node_operations_test.py::RandomNodeOperationsTest.test_node_operations@{"cloud_storage_type":2,"compaction_mode":"chunked_sliding_window","enable_failures":true,"mixed_versions":false,"with_iceberg":false}
```
<!--/body-->
